modules/apps/portal-remote/portal-remote-cors-impl/src/main/java/com/liferay/portal/remote/cors/internal/DynamicURLToCORSSupportMapper.java

```diff
@@ -0,0 +1,269 @@
+				TrieNode nextTrieNode = currentTrieNode.next('*');
+
+				if ((nextTrieNode != null) && nextTrieNode.isEnd()) {
+					corsSupport = nextTrieNode.get();
+				}
+			}
+
+			previousTrieNode = currentTrieNode;
+		}
+
+		if (currentTrieNode != null) {
+			if (onlyExact) {
+				if (!currentTrieNode.isEnd()) {
+					return null;
+				}
+
+				return currentTrieNode.get();
+			}
+
+			if (!onlyWildcard && currentTrieNode.isEnd()) {
+				return currentTrieNode.get();
+			}
+
+			currentTrieNode = currentTrieNode.next('/');
+
+			if (currentTrieNode != null) {
+				currentTrieNode = currentTrieNode.next('*');
+
+				if ((currentTrieNode != null) && currentTrieNode.isEnd()) {
+					corsSupport = currentTrieNode.get();
+				}
+			}
+		}
+
+		return corsSupport;
+	}
+
+	@Override
+	protected void put(
+		CORSSupport corsSupport, String urlPattern, boolean wildcard) {
+
+		TrieNode previousTrieNode = null;
+
+		if (wildcard) {
+			previousTrieNode = _wildCardTrieNode;
+		}
+		else {
+			previousTrieNode = _extensionTrieNode;
+		}
+
+		TrieNode currentTrieNode = null;
+
+		for (int i = 0; i < urlPattern.length(); ++i) {
+			int index = i;
+
+			if (!wildcard) {
+				index = urlPattern.length() - 1 - i;
+			}
+
+			currentTrieNode = previousTrieNode.next(urlPattern.charAt(index));
+
+			if (currentTrieNode == null) {
+				TrieNode nextNode = _trieNodeHeap.nextAvailableNode();
+
+				currentTrieNode = previousTrieNode.setNext(
+					urlPattern.charAt(index), nextNode);
+			}
+
+			previousTrieNode = currentTrieNode;
+		}
+
+		if (currentTrieNode != null) {
+			currentTrieNode.set(corsSupport);
+		}
+	}
+
+	private final TrieNode _extensionTrieNode;
+	private final TrieNodeHeap _trieNodeHeap;
+	private final TrieNode _wildCardTrieNode;
+
+	private static class TrieNode {
+
+		public TrieNode() {
+			_trieNodes = new ArrayList<>(ASCII_CHARACTER_RANGE);
+
+			for (int i = 0; i < ASCII_CHARACTER_RANGE; ++i) {
+				_trieNodes.add(null);
+			}
+		}
+
+		public CORSSupport get() {
+			return _corsSupport;
+		}
+
+		public boolean isEnd() {
+			if (_corsSupport != null) {
+				return true;
+			}
+
+			return false;
+		}
+
+		public TrieNode next(char character) {
+			int index = character - ASCII_PRINTABLE_OFFSET;
+
+			if ((index < 0) ||
+				(index >= (ASCII_CHARACTER_RANGE + ASCII_PRINTABLE_OFFSET))) {
+
+				throw new IllegalArgumentException();
+			}
+
+			return _trieNodes.get(index);
+		}
+
+		public void set(CORSSupport corsSupport) {
+			_corsSupport = corsSupport;
+		}
+
+		public TrieNode setNext(char character, TrieNode nextNode) {
+			int index = character - ASCII_PRINTABLE_OFFSET;
+
+			if ((index < 0) ||
+				(index >= (ASCII_CHARACTER_RANGE + ASCII_PRINTABLE_OFFSET))) {
+
+				throw new IllegalArgumentException();
+			}
+
+			_trieNodes.set(index, nextNode);
+
+			return nextNode;
+		}
+
+		private CORSSupport _corsSupport;
+		private final List<TrieNode> _trieNodes;
+
+	}
+
+	private static class TrieNodeHeap {
```

Comment:
I think we should rename this to `TrieNodeArrayList`

Reply:
I thought heap better presents the idea it's a heap for nodes?

Reply:
yeah... for me it is clear... but we normally name the things after the types it contains... I am fine to call it heap but maybe we should make a point to Brian about it